### ISSUE TITLE
Wrapping SQLException must copy SQLState and other info from wrapped SQLException

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#14696]
The new feature https://github.com/jOOQ/jOOQ/issues/14675 has exposed a bug that hasn't been seen so far.

Some JDBC drivers already throw `DataTruncation` on the `PreparedStatement::setString` call, e.g. Firebird:

```java
try {
    connection.createStatement().execute("create table t (i varchar(3))");

    try (PreparedStatement p = connection.prepareStatement("insert into t values (?)")) {
        try {
            p.setString(1, "abcd"); // This throws DataTruncation
        }
        catch (SQLException e) {
            System.out.println(e.getSQLState());
        }
    }
}
finally {
    connection.createStatement().execute("drop table t");
}
```

jOOQ's `DefaultBindContext` then wraps this in a new `SQLException`, adding some additional information:

```java
try {
    ((Field<Object>) field).getBinding().set(
        new DefaultBindingSetStatementContext<>(new SimpleExecuteContext(configuration(), data()), stmt, nextIndex, value)
    );
}
catch (Exception e) {
    throw new SQLException("Error while writing value at JDBC bind index: " + nextIndex ,e);
}
```

This removes access to the underlying `SQLException::getSQLState`, and possibly other information. The intention was probably to wrap `RuntimeException` and others, but not `SQLException`. The latter should be passed through, thus:

```java
try {
    ((Field<Object>) field).getBinding().set(
        new DefaultBindingSetStatementContext<>(new SimpleExecuteContext(configuration(), data()), stmt, nextIndex, value)
    );
}
catch (SQLException e) {
    throw e;
}
catch (Exception e) {
    throw new SQLException("Error while writing value at JDBC bind index: " + nextIndex ,e);
}
```

**Comment by lukaseder** [comment]
The reason for the wrapping is: https://github.com/jOOQ/jOOQ/issues/5901. The `SQLException` might have already been thrown by jOOQ itself. So we must continue to wrap, but copy `SQLState` and other info.

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.18.0, 3.17.9 (https://github.com/jOOQ/jOOQ/issues/14697), and 3.16.15 (https://github.com/jOOQ/jOOQ/issues/14698)
